Large file load

ItIsJordan opened the pull request:
Adds functionality to check tables and additional resources to be checked against a size threshold to restrict/defer loading.

Closes #136.

ItIsJordan committed:
Add record file size check function

Adds a function/web route to check a specific file for whether or not it breaches the size threshold for immediate loading.

ItIsJordan committed:
Add file size check to records JS (wip)

Adds the file size check function to the records JS page, but does not currently act on the response.

ItIsJordan committed:
Revert "Add file size check to records JS (wip)"

This reverts commit e389a94ac5141d946fdc9340eaf628d5ab9417d4.

ItIsJordan committed:
Merge branch 'main' into large_file_load

ItIsJordan committed:
Remove file_size_check function

Removes the previously added file_size_check function as another appproach is to be taken

ItIsJordan committed:
Update get_table_details and add size check

Updates the get_table_details function and adds  a file size loading threshold check

ItIsJordan committed:
Merge branch 'main' into large_file_load

ItIsJordan committed:
Update javascript/html to support staggered loading

Update javascript/html to support staggered loading of the table data through use of a confirmation button before attempting load of a large file.

ItIsJordan committed:
Remove the load_fail value in table_contents

Removes the load_fail value in table_contents due to finding a better implementation. Table data existence is checked by length of "values" in table_data.

ItIsJordan committed:
Size display changes

Improves the filesize display and now displays the size of file after loading failture.

ItIsJordan committed:
Fix js selectors

Fix a few broken selectors in javascript managing the display of the filesize loading button section

ItIsJordan committed:
Table size display update (to MB)

Updates the table size display to convert to megabytes.

ItIsJordan committed:
Fix records_test.py

Updates records_test.py to work following an addition (size) to the table_contents dictionary in the get_table_details function

ItIsJordan committed:
Set size check threshold to megabyte values

Updates the size check threshold in config.py to be easily set as a megabyte value.

ItIsJordan committed:
Add blank submission test helper function

Adds a helper function to create a blank test submission in test setups.

ItIsJordan committed:
Update related record e2e test

Updates the related record test to use the new helper function, and improves the url checking in the test.

ItIsJordan committed:
Add new test record creation helper function

Adds a function to create a dummy record alongside data. Updates a test to use this.

ItIsJordan committed:
Remove unused imports

Removes some unused imports

ItIsJordan committed:
e2e tests for large file loading

Adds a test and data for large file loading functionality

ItIsJordan committed:
Update commenting

Updates commenting for new filesize feature

ItIsJordan committed:
Merge branch 'main' into large_file_load

ItIsJordan committed:
Clear figure on table switch

Adds a line in hepdata_tables.js code upon table switch to clear the figure section.

ItIsJordan committed:
Additional resource threshold check and test

Adds a filesize threshold check to the loading of additional resources on the landing page. Also updates test/adds data for this.

ItIsJordan committed:
Merge branch 'main' into large_file_load

ItIsJordan committed:
Update  additional resource threshold text

Updates the filesize threshold fail html text for additional resources to read less like an error message.

ItIsJordan committed:
Cleanup/commenting

Minor cleanup and commenting work to large file loading changes

GraemeWatt commented on hepdata/config.py:
```diff
@@ -178,6 +178,8 @@ def _(x):
 CONVERT_MAX_SIZE = sys.maxsize  # Limit on payload sent to converter (checked at submission)
 CLIENT_TIMEOUT = 298  # Client-side timeout in s (should be slightly smaller than server timeout)
+SIZE_LOAD_CHECK_THRESHOLD = 1 * (1024 * 1024) # Size threshold for immediate loading of a table on the records page.
+ADDITIONAL_SIZE_LOAD_CHECK_THRESHOLD = 1 * (1024 * 1024) # Size threshold for disallowing render of additional res files
```
Comment that these values are in bytes.

GraemeWatt commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -95,9 +126,45 @@ HEPDATA.table_renderer = {
+        $("#figures").html('');
+
+        // If there is no table, the table is too big
+        if(table_data.values.length == 0) {
```
There are some legitimate data tables that are empty (for example, first table of https://www.hepdata.net/record/ins2719822), so you'll need to work out an alternative way of deciding if the table is too big.  Can't the table `size` be used instead?

GraemeWatt commented on hepdata/modules/records/templates/hepdata_records/components/resource_details.html:
```diff
@@ -34,7 +34,14 @@ <h4>
+          <textarea id="code-contents" height="70%">{{ ctx.contents }}</textarea>
+        {% else %}
+        <div id="code-contents-fail" height="70%">
+          <p style="font-weight:bolder">This file is larger than our loading threshold, and is only available for download below.</p>
```
Can the file size (in MB) and the loading threshold (in MB) be displayed?  For example:

"This file (1.6 MB) is larger than our loading threshold (1.0 MB), and is only available for download below."

GraemeWatt reviewed (changes requested):
Looks good, but I've added a couple of comments.

GraemeWatt commented on hepdata/modules/records/views.py:
```diff
@@ -699,7 +713,13 @@ def get_resource(resource_id):
+                        if mimetypes.guess_type(resource_obj.file_location)[0] != 'application/x-tar':
+                            # Check against the filesize threshold. Do not set contents if it fails.
+                            filesize = os.path.getsize(resource_obj.file_location)
+                            if filesize < ADDITIONAL_SIZE_LOAD_CHECK_THRESHOLD:
+                                contents = resource_file.read()
```
For a large `filesize`, the `contents = ''` (an empty string).  Maybe change to setting `contents = 'Large text file'`, which would be more informative if viewing the JSON format of the landing page.  The code in `resource_details.html` would then need to be modified.

ItIsJordan committed:
Update config commenting

Update commenting of size thresholds in config.py

ItIsJordan committed:
Update additional resource view response

Updates additional resource response to note large file. Updates html to work with this and still display correctly.

ItIsJordan committed:
Add filesize to additional resource page

Displays filesize on the html of the resource_details page when the loading size threshold is broken.

ItIsJordan committed:
Update table renderer check

Updates the table rendering function in hepdata_tables.js to check against both the size and length of the data to ensure empty files (which will still have a size) are checked for

ItIsJordan committed:
Move filesize table render check

Moves the filesize check to run the table data render function in hepdata_tables.js into the function itself. Allowing lines that were skipped to now run.

ItIsJordan committed:
Increase wait times in large file load e2e testing

ItIsJordan committed:
Fix empty data table in large file load testing

Fixes a broken test as the behaviour of empty tables is different than when uploaded

ItIsJordan committed:
Merge branch 'main' into large_file_load

GraemeWatt commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -95,57 +126,95 @@ HEPDATA.table_renderer = {
+        $("#figures").html('');
+
+        // Check that the table is both empty, and is larger than an empty table (bytes)
+        if(table_data.values.length == 0 && table_data.size > 300) {
```
The value of 300 bytes in the condition `table_data.size > 300` seems rather arbitrary.  For the example I gave of an empty table (first table of [ins2719822](https://www.hepdata.net/record/ins2719822)), the size is 125 bytes, but the size will depend on the headers and qualifiers.  An empty table could still have size greater than 300 bytes if there are lengthy headers and qualifiers.  I don't see why you can't directly compare the table size with the `SIZE_LOAD_CHECK_THRESHOLD` config variable by first propagating it to the Javascript in [`record-javascript.html`](https://github.com/HEPData/hepdata/blob/main/hepdata/modules/records/templates/hepdata_records/components/record-javascript.html) like:
```javascript
HEPDATA.size_load_check_threshold = {{ config['SIZE_LOAD_CHECK_THRESHOLD']|default(1*1024*1024) }};
```
then replacing the `if` condition with:
```javascript
if (table_data.size > HEPDATA.size_load_check_threshold) {...}
```

GraemeWatt commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -95,57 +126,95 @@ HEPDATA.table_renderer = {
-          });
-          HEPDATA.table_renderer.attach_row_listener(table_placement, 'histogram');
+        // If it is larger than an empty table (bytes)
+        if(table_data.size > 300) {
```
The headers and qualifiers should still be displayed for an empty table, so this `if` condition should be removed to recover the previous behaviour.

GraemeWatt commented on hepdata/modules/records/views.py:
```diff
@@ -744,6 +767,7 @@ def get_resource(resource_id):
                     status_code = 404 if 'error' in ctx['json_ld'] else 200
                     return jsonify(ctx['json_ld']), status_code
                 else:
+                    if filesize: ctx['filesize'] = '%.2f'%((filesize / 1024) / 1024) # Set filesize if exists
```
Here you can also add the threshold to the `ctx` so it can be used in the HTML message:
```python
ctx['ADDITIONAL_SIZE_LOAD_CHECK_THRESHOLD'] = '%.2f'%((ADDITIONAL_SIZE_LOAD_CHECK_THRESHOLD / 1024) / 1024)
```

GraemeWatt commented on hepdata/modules/records/templates/hepdata_records/components/resource_details.html:
```diff
@@ -34,7 +34,14 @@ <h4>
+          <textarea id="code-contents" height="70%">{{ ctx.contents }}</textarea>
+        {% else %}
+        <div id="code-contents-fail" height="70%">
+          <p style="font-weight:bolder">This file ({{ ctx.filesize }}MB) is larger than our loading threshold (1.0MB), and is only available for download below.</p>
```
Better to avoid hard-coding the threshold ("1.0MB") and add a space before "MB" (twice):
```
This file ({{ ctx.filesize }} MB) is larger than our loading threshold ({{ ctx.ADDITIONAL_SIZE_LOAD_CHECK_THRESHOLD }} MB)
```

ItIsJordan committed:
Update additional resource load threshold html

Updates the additional resource size threshold in resource_details.html to support formatted display of the current threshold. Also improves current text.

ItIsJordan committed:
Update record-javascript.html

Add size load check threshold to the javascript in record-javascript.html

ItIsJordan committed:
Update hepdata_tables.js render code

Updates the code managing the table data rendering to revert to previous behaviour of always showing headers etc. Improves the table_data size check.

ItIsJordan committed:
Update records threshold display

Updates the records page size threshold error display to also show the threshold size.

sjmf commented on hepdata/modules/dashboard/assets/js/hepdata_dashboard_submissions_vis.js:
```diff
@@ -71,7 +71,6 @@ var submissions_vis = (function () {
   }
 
   function resetCharts(charts) {
-    console.log("Aaargh");
```
😂👍

sjmf commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -15,7 +15,17 @@ HEPDATA.switch_table = function (listId, table_requested, table_name, status) {
+      {x: -27, y: -30, color: "#955BA5"},
+      {x: 60, y: -55, color: "#2C3E50"}],
+    {"width": 200, "height": 200}
+  );
```
Just a thought: if this is how it is done elsewhere in the HEPData codebase then it's fine to do it here too, but generally I would have thought applying a class and handling styling in CSS to be the better approach.

sjmf replied:
Oh, unless this is calling a d3 function? I guess it might be difficult to do that with CSS...

sjmf commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -95,17 +126,55 @@ HEPDATA.table_renderer = {
+          $("#hepdata_table_loader").addClass("hidden");
+          var megabyte_size = (table_data.size / (1024 * 1024)).toFixed(2);
+          var additional_size = (HEPDATA.size_load_check_threshold / (1024 * 1024)).toFixed(2);
+          var message = "Table size is: " + megabyte_size + " MB, which is larger than our threshold of " + additional_size + " MB.";
```
I don't know if I like the idea of setting messages directly in Javascript. I'd prefer to show/hide messages that already exist in the HTML using `<span class="megabyte_size"></span>` fields to set the values.

The principle here is separation of concerns. HTML is where the markup lives. CSS is where the styling lives. JS is for control and interactivity.

sjmf commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -95,17 +126,55 @@ HEPDATA.table_renderer = {
+      },
+      error: function (data, error) {
+        console.error(error);
+        d3.select("#hepdata_table_loading_failed_text").html('Failed to load table defined by ' + url);
```
The other approach could be to have a library of possible messages to reduce the tendency towards magic strings. The current approach would make it very difficult if HEPData ever needs to be localized to other languages! But, if this is the way things are currently done, then stick with the existing standard.

sjmf commented on hepdata/modules/records/utils/common.py:
```diff
@@ -251,3 +251,19 @@ def get_record_by_id(recid):
+    # We do the check only if told to, otherwise we just pass as true
+    if load_all == 0:
+        size_check["status"] = size_check["size"] <= SIZE_LOAD_CHECK_THRESHOLD
+    return size_check
```
Good 👍

sjmf commented on tests/test_data/TestLargeSubmission/additional1.yaml:
I don't know if committing a 65MB file into git was a good idea, but it's in the index now! Just means the size of the repo download will grow. Hopefully it compresses well.

sjmf replied:
I've not used [git-lfs ](https://github.com/git-lfs/git-lfs) but that may be an alternative in future.

GraemeWatt replied:
The file is only 1.12 MB, although it has 65020 lines.

sjmf replied:
My mistake! This is fine.

ItIsJordan committed:
Update filesize html messaging

Modifies functionality of javascript used to display table size on the records page when the threshold is broken to move main message content to html.

ItIsJordan committed:
Update large file load test

Updates the large file load testing to update regex patterns to current.

GraemeWatt committed:
Merge branch 'main' into large_file_load
